website_sale_product_assortment/controllers/website_sale.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2021 Tecnativa - Carlos Roca
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from werkzeug.exceptions import NotFound
+
+from odoo.http import request, route
+from odoo.osv import expression
+
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+
+
+class WebsiteSale(WebsiteSale):
+    def _get_products_allowed(self):
+        partner = request.env.user.partner_id
+        assortments = (
+            request.env["ir.filters"]
+            .sudo()
+            .search(
+                [
+                    ("is_assortment", "=", True),
+                    ("website_availability", "=", "no_show"),
+                ]
+            )
+        )
+        assortment_restriction = False
+        allowed_products = request.env["product.product"]
+        for assortment in assortments:
+            if (
+                # Set active_test to False to allow filtering by partners
+                # that are not active, (for example Public User)
+                partner
+                & assortment.sudo().with_context(active_test=False).all_partner_ids
+            ):
+                assortment_restriction = True
+                allowed_products |= request.env["product.product"].search(
+                    assortment._get_eval_domain()
+                )
+        return allowed_products, assortment_restriction
+
+    @route()
+    def product(self, product, category="", search="", **kwargs):
+        """Overriding product method to avoid accessing to product sheet when the
+        product assortments prevent to show them.
+        """
+        allowed_products, assortment_restriction = self._get_products_allowed()
+        if assortment_restriction:
+            not_allowed_products = request.env["product.product"].search(
+                [("id", "not in", allowed_products.ids)]
+            )
+            if len(product.product_variant_ids & not_allowed_products) == len(
+                product.product_variant_ids
+            ):
+                raise NotFound()
+        return super().product(product, category=category, search=search, **kwargs)
+
+    def _get_search_domain(
+        self, search, category, attrib_values, search_in_description=True
+    ):
+        """Overriding _get_search_domain method to avoid show product templates that
+        has all their variants not allowed to be shown.
+        """
+        res = super()._get_search_domain(
+            search, category, attrib_values, search_in_description=search_in_description
+        )
+        allowed_product_variants, assortment_restriction = self._get_products_allowed()
```

Comment:
Can you avoid calling 2 times to the same method?

Reply:
I don't know were it is called 2 times... There are 2 different methods that call `_get_products_allowed()`

Reply:
Yes, in this file you can see several calls to the same method, and you should do it only once for performance reasons.